File: LXUtilities/MBProgressHUDExtensions/MBProgressHUD+LXExtension.m
//
//  MBProgressHUD+LXExtension.m
//
//  Created by 从今以后 on 15/9/12.
//  Copyright © 2015年 从今以后. All rights reserved.
//

#import "LXUtilities.h"
#import "LXRingAnimatedView.h"
#import "MBProgressHUD+LXExtension.h"

NS_ASSUME_NONNULL_BEGIN

@implementation MBProgressHUD (LXExtension)

#pragma mark - 短暂显示自定义图标 HUD

+ (MBProgressHUD *)lx_showStatus:(NSString *)status withImage:(UIImage *)image toView:(UIView *)view
{
    NSParameterAssert(view);
    NSParameterAssert(image);
    MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:view animated:YES];
    hud.label.text = status;
    hud.mode = MBProgressHUDModeCustomView;
    hud.customView = [[UIImageView alloc] initWithImage:image];
    [hud hideAnimated:YES afterDelay:1.0];
    return hud;
}

#pragma mark - 短暂显示纯文本 HUD

+ (MBProgressHUD *)lx_showStatus:(NSString *)status toView:(nullable UIView *)view
{
    NSParameterAssert(status.length > 0);
    MBProgressHUD *hud = [self showHUDAddedTo:view ?: [UIWindow lx_topWindow] animated:YES];
    hud.label.text = status;
    hud.mode = MBProgressHUDModeText;
    [hud hideAnimated:YES afterDelay:1.0];
    return hud;
}

#pragma mark - 短暂显示带图标的文本

+ (MBProgressHUD *)lx_showInfoWithStatus:(NSString *)status toView:(nullable UIView *)view {
    return [self lx_showStatus:status withImage:[UIImage imageNamed:@"info"] toView:view ?: [UIWindow lx_topWindow]];
}

+ (MBProgressHUD *)lx_showSuccessWithStatus:(NSString *)status toView:(nullable UIView *)view {
    return [self lx_showStatus:status withImage:[UIImage imageNamed:@"success"] toView:view ?: [UIWindow lx_topWindow]];
}

+ (MBProgressHUD *)lx_showFailureWithStatus:(NSString *)status toView:(nullable UIView *)view {
    return [self lx_showStatus:status withImage:[UIImage imageNamed:@"failure"] toView:view ?: [UIWindow lx_topWindow]];
}

#pragma mark - 持续显示环形动画活动指示器

+ (MBProgressHUD *)lx_showRingActivityIndicatorWithStatus:(nullable NSString *)status toView:(nullable UIView *)view
{
    MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:view ?: [UIWindow lx_topWindow] animated:YES];
    hud.label.text = status;
    hud.mode = MBProgressHUDModeCustomView;
    hud.customView = [LXRingAnimatedView new];
    return hud;
}

#pragma mark - 持续显示原生活动指示器

+ (MBProgressHUD *)lx_showActivityIndicatorWithStatus:(nullable NSString *)status toView:(nullable UIView *)view
{
    MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:view ?: [UIWindow lx_topWindow] animated:YES];
    hud.label.text = status;
    return hud;
}

#pragma mark - 持续显示伴有遮罩背景的原生活动指示器

+ (MBProgressHUD *)lx_showMaskActivityIndicatorWithStatus:(nullable NSString *)status toView:(nullable UIView *)view
{
    MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:view ?: [UIWindow lx_topWindow] animated:YES];
    hud.label.text = status;
    hud.backgroundView.style = MBProgressHUDBackgroundStyleSolidColor;
    hud.backgroundView.color = [UIColor colorWithWhite:0.f alpha:.2f];
    return hud;
}

#pragma mark - 持续显示环形进度条 HUD

+ (MBProgressHUD *)lx_showProgressWithStatus:(nullable NSString *)status toView:(nullable UIView *)view
{
    MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:view ?: [UIWindow lx_topWindow] animated:YES];
    hud.label.text = status;
    hud.mode = MBProgressHUDModeAnnularDeterminate;
    return hud;
}

#pragma mark - 隐藏 HUD

+ (void)lx_hideHUDForWindow {
    [self lx_hideHUDForWindowAnimated:YES];
}

+ (void)lx_hideHUDForWindowAnimated:(BOOL)animated {
    [self hideHUDForView:[UIWindow lx_topWindow] animated:animated];
}

- (void)lx_hideAnimated:(BOOL)animated afterDelay:(NSTimeInterval)delay {
    NSTimer *timer = [self valueForKey:@"hideDelayTimer"];
    if (timer) {
        [timer invalidate];
    }
    [self hideAnimated:animated afterDelay:delay];
}

@end

NS_ASSUME_NONNULL_END

File: LXUtilities/MBProgressHUDExtensions/MBProgressHUD+LXExtension.swift
//
//  MBProgressHUD+LXExtension.swift
//
//  Created by 从今以后 on 2017/11/30.
//  Copyright © 2017年 从今以后. All rights reserved.
//

import MBProgressHUD

extension Swifty where Base: MBProgressHUD {

    /// 创建 HUD 并添加到指定视图上，持续显示原生风格的活动指示器。
    ///
    /// - Parameter view: HUD 的父视图，默认为顶层窗口
    /// - Returns:        创建的 HUD 实例
    @discardableResult static func showIndicator(to view: UIView = UIWindow.lx.topWindow()) -> MBProgressHUD {
        let hud = MBProgressHUD.showAdded(to: view, animated: true)

        hud.margin = 16
        hud.mode = .indeterminate
        hud.bezelView.color = .clear
        hud.bezelView.style = .solidColor

        return hud
    }

    /// 创建 HUD 并添加到指定视图上，短暂显示提示文字后自动移除。
    ///
    /// - Parameters:
    ///   - status: 提示文字
    ///   - view:   HUD 的父视图，默认为顶层窗口
    /// - Returns:  创建的 HUD 实例
    @discardableResult static func showStatus(_ status: String, to view: UIView = UIWindow.lx.topWindow()) -> MBProgressHUD {
        let hud = MBProgressHUD.showAdded(to: view, animated: true)

        hud.margin = 16
        hud.mode = .text
        hud.label.text = status
        hud.contentColor = .white
        hud.bezelView.style = .solidColor
        hud.bezelView.color = UIColor(white: 0, alpha: 0.7)

        hud.hide(animated: true, afterDelay: 1)

        return hud
    }


    /// 创建水平进度条 HUD 并添加到指定视图上。
    ///
    /// - Parameters:
    ///   - status: 提示文字
    ///   - view:   HUD 的父视图，默认为顶层窗口
    /// - Returns:  创建的 HUD 实例
    @discardableResult static func showProgressBar(withStatus status: String, to view: UIView = UIWindow.lx.topWindow()) -> MBProgressHUD {
        let hud = MBProgressHUD.showAdded(to: view, animated: true)

        hud.margin = 16
        hud.label.text = status
        hud.contentColor = .white
        hud.bezelView.style = .solidColor
        hud.mode = .determinateHorizontalBar
        hud.bezelView.color = UIColor(white: 0, alpha: 0.7)

        return hud
    }

	/// 将当前 HUD 转为持续显示原生风格的活动指示器的 HUD。
	///
	/// - Returns: 当前 HUD 实例
	@discardableResult func showIndicator() -> MBProgressHUD {
        base.label.text = nil
        base.mode = .indeterminate
        base.bezelView.color = .clear
        base.bezelView.style = .solidColor
        base.contentColor = UIColor(white: 0, alpha: 0.7)

        base.superview?.bringSubview(toFront: base)

		return base
	}

    /// 将当前 HUD 转为短暂显示提示文字的 HUD 并在一秒后自动移除。
    ///
    /// - Parameters:
    ///   - status: 提示文字
    /// - Returns:  当前 HUD 实例
    @discardableResult func showStatus(_ status: String) -> MBProgressHUD {
        base.mode = .text
        base.label.text = status
        base.contentColor = .white
        base.bezelView.style = .solidColor
        base.bezelView.color = UIColor(white: 0, alpha: 0.7)

        base.hide(animated: true, afterDelay: 1)
        base.superview?.bringSubview(toFront: base)

        return base
    }
}
